blocks_vertical/procedures.js

```diff
@@ -312,91 +345,208 @@ Blockly.ScratchBlocks.ProcedureUtils.attachArgumentReporter_ = function(
- * @param {string} name The name to use when adding the input to the block.
  * @param {number} index The index of this input into the argument id array.
  * @param {!Object.<string, Blockly.Block>} connectionMap An object mapping
  *     parameter IDs to the blocks that were connected to those IDs at the
  *     beginning of the mutation.
+ * @param {Blockly.BlockSvg} oldBlock The block that was previously connected to
+ *     this input, or null.
+ * @param {!Blockly.Input} input The newly created input to populate.
  * @private
  * @this Blockly.Block
  */
-Blockly.ScratchBlocks.ProcedureUtils.createInput_ = function(type, name,
-    index, connectionMap) {
-  var id = this.argumentIds_[index];
-  var oldBlock = null;
-  if (connectionMap && (id in connectionMap)) {
-    oldBlock = connectionMap[id];
-  }
-  var input = this.appendValueInput(name);
+Blockly.ScratchBlocks.ProcedureUtils.populateInputCaller_ = function(type, index,
+    connectionMap, oldBlock, input) {
   if (connectionMap && oldBlock) {
-    this.reattachBlock_(input, type, oldBlock, id, connectionMap);
+    // Reattach the old block.
+    connectionMap[input.name] = null;
+    oldBlock.outputConnection.connect(input.connection);
+    if (type != 'b') {
+      input.connection.setShadowDom(this.buildShadowDom_(type));
+    }
   } else {
     this.attachShadow_(input, type);
   }
-  this.paramMap_[id] = input;
 };
 
 /**
- * Create an input, attach the correct block to it, and insert it into the
- * params map.
+ * Populate the given input with the correct child block or values.
  * This function is used by the procedures_callnoreturn_internal block.
  * TODO (#1213) consider renaming.
  * @param {string} type One of 'b' (boolean), 's' (string) or 'n' (number).
- * @param {string} name The name to use when adding the input to the block.
  * @param {number} index The index of this input into the argument id and name
  *     arrays.
  * @param {!Object.<string, Blockly.Block>} connectionMap An object mapping
  *     parameter IDs to the blocks that were connected to those IDs at the
  *     beginning of the mutation.
+ * @param {Blockly.BlockSvg} oldBlock The block that was previously connected to
+ *     this input, or null.
+ * @param {!Blockly.Input} input The newly created input to populate.
  * @private
  * @this Blockly.Block
  */
-Blockly.ScratchBlocks.ProcedureUtils.createInputCallerInternal_ = function(type,
-    name, index, connectionMap) {
-  var id = this.argumentIds_[index];
-  var oldBlock = null;
-  if (connectionMap && (id in connectionMap)) {
-    oldBlock = connectionMap[id];
+Blockly.ScratchBlocks.ProcedureUtils.populateInputCallerInternal_ = function(type,
+    index, connectionMap, oldBlock, input) {
+  var oldTypeMatches =
+    Blockly.ScratchBlocks.ProcedureUtils.checkOldTypeMatches_(oldBlock, type);
+  var displayName = this.displayNames_[index];
+  if (connectionMap && oldBlock && oldTypeMatches) {
```

Comment:
Not important but when oldBlock is null, oldTypeMatches returns false so you could get rid of it in the condition